### PR TITLE
feat: Show Progress in Taskbar/Dock

### PR DESCRIPTION
**Pull request by NightFeather0615**
#166 
Windows side should be done, I'm currently trying to implement it on macOS.

### CODE REVIEW

**Comment by NightFeather0615** [comment]
Well, there doesn't seem to be a package for macOS, so I wrote one for it. [Here's the repo](https://github.com/NightFeather0615/macos_dock_progress).

Everything should be ready to merge now! Please let me know if there is a way to write tests for this feature if they are needed.

**Comment by Tienisto** [comment]
Thank you for the pull request and sorry for the delay.

I think you have written lots of if-else branches there. It's better to encapsualte this logic into one or many util functions.

I think `lib/util/native` would be a great place for it: https://github.com/localsend/localsend/tree/main/app/lib/util/native

File: app/lib/pages/progress_page.dart [hunk 1]
```diff
@@ -114,6 +117,11 @@ class _ProgressPageState extends State<ProgressPage> with Refena {
   void dispose() {
     super.dispose();
     _finishTimer?.cancel();
+    if (Platform.isWindows) {
```
Comment (Tienisto) [inline app/lib/pages/progress_page.dart:120]: Extract platform-specific handling to utils

File: app/lib/pages/progress_page.dart [hunk 2]
```diff
@@ -162,6 +170,18 @@ class _ProgressPageState extends State<ProgressPage> with Refena {
     final currBytes = _files.fold<int>(
         0, (prev, curr) => prev + ((progressNotifier.getProgress(sessionId: widget.sessionId, fileId: curr.id) * curr.size).round()));
 
+    if (_totalBytes != double.minPositive.toInt() && _totalBytes != double.maxFinite.toInt()) {
+      if (Platform.isWindows) {
```
Comment (Tienisto) [inline app/lib/pages/progress_page.dart:174]: Extract platform-specific handling to utils

File: app/lib/pages/receive_page.dart [hunk 3]
```diff
@@ -41,6 +43,14 @@ class _ReceivePageState extends State<ReceivePage> with Refena {
     WidgetsBinding.instance.addPostFrameCallback((_) async => _init());
   }
 
+  @override
+  void dispose() {
+    super.dispose();
+    if (Platform.isWindows) {
```
Comment (Tienisto) [inline app/lib/pages/receive_page.dart:49]: Extract platform-specific handling to utils (here you can do a noop for Linux (?) and macOS)

File: app/lib/pages/send_page.dart [hunk 4]
```diff
@@ -33,6 +37,14 @@ class _SendPageState extends State<SendPage> with Refena {
   Device? _myDevice;
   Device? _targetDevice;
 
+  @override
+  void dispose() {
+    super.dispose();
+    if (Platform.isWindows) {
```
Comment (Tienisto) [inline app/lib/pages/send_page.dart:43]: Extract platform-specific handling to utils (here you can do a noop for Linux (?) and macOS)

**Comment by NightFeather0615** [comment]
Thanks for the review!
Since the code still looks bad after extracting platform-specific handlers into the utils, I've wrapped the taskbar/dock-related functions into a helper class, hopefully this will make the code a bit cleaner.

**Comment by Tienisto** [comment]
Thank you!

**Pull request merged**